### PR TITLE
audio attachments

### PR DESCRIPTION
**Pull request by mmaxim**
Exploration/initial implementation of audio attachments. I tried to get as close to the design as possible, but it is likely not exact. The hope was to prove that the whole thing could work on iOS/Android, and I believe this does it. Here is a demo of how it looks on my phone: 

audio

**Note, this is not a finished product, and is behind a feature flag (which is currently set to on, so people can play with this)**

Things not done (for later work):

1.) Send waveform display
2.) Preview clip before sending after lock mode
3.) Metering on Android

cc @adamjspooner @cecileboucheron @malgorithms

### CODE REVIEW

**Comment by malgorithms** [comment]
@mmaxim - what's your thought on this one:

> Metering on Android

Is that pretty doable? 

I have no doubt about the other 2 items on your list.

**Comment by mmaxim** [comment]
> Is that pretty doable?

Very doable, I just didn't do it because I wanted to get to a stopping point.

File: shared/actions/platform-specific/index.native.tsx [hunk 1]
```diff
@@ -683,6 +705,76 @@ const configureFileAttachmentDownloadForAndroid = () =>
...
+    AudioQuality: 'Low',
+    Channels: 1,
+    MeteringEnabled: true,
+    SampleRate: 22050,
+  })
+  AudioRecorder.onProgress = data => {
+    action.payload.meteringCb(data.currentMetering)
+  }
+  AudioRecorder.onFinished = data => {
+    logger.info('startAudioRecording: recording finished: ' + JSON.stringify(data))
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.native.tsx]: 🔥 lets not leak this

File: shared/actions/platform-specific/index.native.tsx [hunk 2]
```diff
@@ -683,6 +705,76 @@ const configureFileAttachmentDownloadForAndroid = () =>
...
+    if (audio && ChatConstants.showAudioRecording(audio) && audio.isLocked) {
+      return false
+    }
+  }
+  logger.info('stopAudioRecording: stopping recording')
+  try {
+    await AudioRecorder.stopRecording()
+  } catch (e) {}
+
+  if (!state.chat2.audioRecording) {
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.native.tsx]: this isn't safe. lets discuss

File: shared/android/app/build.gradle [hunk 3]
```diff
@@ -230,6 +230,7 @@ play {
 }
 
 dependencies {
+    implementation project(':react-native-audio')
```
Comment (chrisnojima) [inline shared/android/app/build.gradle]: shouldn't be needed w/ auto link?

File: shared/android/settings.gradle [hunk 4]
```diff
@@ -1,4 +1,6 @@
 rootProject.name = 'Keybase'
+include ':react-native-audio'
```
Comment (chrisnojima) [inline shared/android/settings.gradle]: ditto ^

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx [hunk 5]
```diff
@@ -171,32 +174,38 @@ class _PlatformInput extends PureComponent<PlatformInputPropsInternal, State> {
...
-            // change to match desktop.
-            onChangeText={this._onChangeText}
-            onSelectionChange={this.props.onSelectionChange}
-            ref={this._inputSetRef}
-            style={styles.input}
-            textType="Body"
-            rowsMax={Styles.dimensionHeight < 600 ? 5 : 9}
-            rowsMin={1}
-          />
+          {
```
Comment (chrisnojima) [inline shared/chat/conversation/input-area/normal/platform-input.native.tsx]: ???

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx [hunk 6]
```diff
@@ -221,9 +234,19 @@ type ActionProps = {
 }
 
 const Action = React.memo(
-  ({hasText, insertMentionMarker, isEditing, onSubmit, openFilePicker, openMoreMenu}: ActionProps) => {
+  ({
```
Comment (chrisnojima) [inline shared/chat/conversation/input-area/normal/platform-input.native.tsx]: lets not explode these things in the param list. can pass as `props` and just pull it out inside the function

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx [hunk 7]
```diff
@@ -291,6 +321,76 @@ const Action = React.memo(
...
+  lockRecording: () => void
+  startRecording: () => void
+  stopRecording: (st: Types.AudioStopType) => void
+}
+
+const maxAudioDrift = -20
+
+const AudioStarter = (props: AudioStarterProps) => {
+  let longPressTimer
+  return flags.audioAttachments ? (
```
Comment (chrisnojima) [inline shared/chat/conversation/input-area/normal/platform-input.native.tsx]: can reduce indents with 
```
if (!flags.audioAttachments) { return ...}
...
```

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx [hunk 8]
```diff
@@ -291,6 +321,76 @@ const Action = React.memo(
...
+    <Kb.TapGestureHandler
+      onHandlerStateChange={({nativeEvent}) => {
+        if (!props.recording && nativeEvent.state === Kb.GestureState.BEGAN) {
+          if (!longPressTimer) {
+            longPressTimer = setTimeout(props.startRecording, 200)
+          }
+        }
+        if (nativeEvent.state === Kb.GestureState.ACTIVE || nativeEvent.state === Kb.GestureState.END) {
+          clearTimeout(longPressTimer)
```
Comment (chrisnojima) [inline shared/chat/conversation/input-area/normal/platform-input.native.tsx]: likely want to clear it out

File: shared/ios/Podfile [hunk 9]
```diff
@@ -35,6 +35,8 @@ target 'Keybase' do
 
   use_unimodules!(modules_paths: ['../node_modules'])
   use_native_modules!
+  pod 'RNAudio', :path => '../node_modules/react-native-audio'
```
Comment (chrisnojima) [inline shared/ios/Podfile]: shouldn't be necessary w/ auto link

File: shared/util/timestamp.tsx [hunk 10]
```diff
@@ -104,6 +104,12 @@ export const formatDuration = (duration: number): string => {
   const d = moment.duration(duration)
   return d.hours() ? `${d.hours()} hr` : d.minutes() ? `${d.minutes()} min` : `${d.seconds()} s`
 }
+
+export const formatAudioRecordDuration = (duration: number): string => {
```
Comment (chrisnojima) [inline shared/util/timestamp.tsx]: should add mock also

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx (hunk 5 above)
Comment (mmaxim) [reply shared/chat/conversation/input-area/normal/platform-input.native.tsx]: Sorry I was hiding this whole thing at one point and just left that in there.

File: shared/chat/conversation/input-area/normal/platform-input.native.tsx [hunk 11]
```diff
@@ -332,7 +329,10 @@ const maxAudioDrift = -20
 
 const AudioStarter = (props: AudioStarterProps) => {
   let longPressTimer
-  return flags.audioAttachments ? (
+  if (!flags.audioAttachments) {
+    return null
```
Comment (chrisnojima) [inline shared/chat/conversation/input-area/normal/platform-input.native.tsx]: return the icon?

**Pull request merged**